File: challenge/202105/week1/01_Prefix_and_Suffix_Search/main.py
# Prefix and Suffix Search
from typing import List


class WordFilter:

    def __init__(self, words: List[str]):
        self.db = dict()

        for i, word in enumerate(words):
            for start in range(len(word), -1, -1):
                for end in range(len(word) + 1):
                    substr = f'{word[start:]}#{word[:end]}'
                    self.db[substr] = i
        
    def f(self, prefix: str, suffix: str) -> int:
        target = f'{suffix}#{prefix}'

        if target in self.db:
            return self.db[target]

        return -1
        

if __name__ == '__main__':
    words = ["apple"]
    prefix, suffix = "a", "e"
    obj = WordFilter(words)
    result = obj.f(prefix, suffix)
    assert result == 0

File: challenge/202105/week1/01_Prefix_and_Suffix_Search/main.go
// Prefix and Suffix Search
package main

import "fmt"

type WordFilter struct {
	db map[string]int
}

func Constructor(words []string) WordFilter {
	db := make(map[string]int)
	substr := ""

	for i, word := range words {
		for s := len(word); s > -1; s-- {
			for e := 0; e < len(word)+1; e++ {
				substr = word[s:] + "#" + word[:e]
				db[substr] = i
			}
		}
	}

	return WordFilter{db}
}

func (this *WordFilter) F(prefix string, suffix string) int {
	target := suffix + "#" + prefix

	if _, ok := this.db[target]; ok {
		return this.db[target]
	} else {
		return -1
	}
}

func print(answer int, expected int) {
	fmt.Printf("Answer:\t\t%d\nExpected:\t%d\n\n", answer, expected)
}

func main() {
	words := []string{"apple"}
	fix := []string{"a", "e"}
	obj := Constructor(words)
	result := obj.F(fix[0], fix[1])
	print(result, 0)
}
